services:
  app:
    build: ./app
    container_name: flask-python_app
    environment:
      - DB_HOST=db
      - DB_USER=${MYSQL_USER}
      - DB_PASSWORD=${MYSQL_PASSWORD}
      - DB_NAME=${MYSQL_DATABASE}
    networks:
      - flask-network
    volumes:
      - ./app:/app
    depends_on:
      db:
        condition: service_healthy 
    command: gunicorn -w 4 -b 0.0.0.0:8000 wsgi:app
  
  web:
    image: nginx:latest
    container_name: flask-nginx_server
    ports:
      - "80:80"
    volumes:
      - ./nginx/nginx.conf:/etc/nginx/nginx.conf:ro
      - ./app:/app
    networks:
      - flask-network
    depends_on:
      - app
  
  db:
    image: mysql:latest
    container_name: flask-mysql_db
    env_file:
      - ./db/.env
    volumes:
      - db_data:/var/lib/mysql
      - ./db/init.sql:/docker-entrypoint-initdb.d/init.sql
    ports:
      - 3306:3306
    networks:
      - flask-network
    healthcheck:
      test: ["CMD", "mysqladmin", "ping", "-h", "localhost"]
      timeout: 10s
      retries: 10

networks:
  flask-network:
    external: true

volumes:
  db_data: